### PR TITLE
feat(shared-tree): Allow setting of value fields on struct nodes

### PR DESCRIPTION
**Pull request by Josmithr**
Also renames type parameters of `FieldSchema` to make their type-parameter-ness more apparent.

[AB#5544](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/5544)

### CODE REVIEW

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts [hunk 1]
```diff
@@ -428,33 +433,42 @@ export type StructTyped<TSchema extends StructSchema> = Struct &
...
-// & {
-// 	// Setter properties (when the type system permits)
-// 	[key in keyof TFields]: UnwrappedField<TFields[key]> & StructSetContent<TFields[key]>;
-// }
+export type StructFields<TFields extends RestrictiveReadonlyRecord<string, FieldSchema>> = {
+	// boxed fields (TODO: maybe remove these when same as non-boxed version?)
+	readonly [key in keyof TFields as `boxed${Capitalize<key & string>}`]: TypedField<TFields[key]>;
+} & {
+	// TODO: readonly uncondionally + setter of intersection type
+	// - Use never as key when not creating setter
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts:450]:
```suggestion
```

File: api-report/tree2.api.md [hunk 2]
```diff
@@ -485,6 +485,9 @@ export interface EditableField extends UntypedField<EditableTreeContext, Editabl
     setContent(newContent: NewFieldContent): void;
 }
 
+// @alpha
+export type EditableFieldKinds = typeof FieldKinds.optional | typeof FieldKinds.required;
```
Comment (taylorsw04) [inline api-report/tree2.api.md:489]: How about `AssignableFieldKinds`? This name is pretty confusing :D

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyTree.ts [hunk 3]
```diff
@@ -516,34 +521,62 @@ function buildStructClass<TSchema extends StructSchema>(
 	const propertyDescriptorMap: PropertyDescriptorMap = {};
 	const ownPropertyMap: PropertyDescriptorMap = {};
 
-	for (const [key, field] of schema.structFields) {
+	for (const [key, fieldSchema] of schema.structFields) {
+		// Map containing setter implementations for field kinds that support setters.
+		const setters = new Map<FieldKind, (newContent: ContextuallyTypedNodeData) => void>([
```
Comment (taylorsw04) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyTree.ts:526]: Considering this is likely a hot path for tree inserts, I think it's worth optimizing this a bit:
- can we avoid making this map and use a ternary instead?
- or, can we initialize it statically and re-bind the methods each time (or pass in the instance)?

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/editableTreeTypes.spec.ts [hunk 4]
```diff
@@ -139,13 +139,13 @@ describe("editableTreeTypes", () => {
 		assert(child);
 
 		// TODO: add shorthand setters
```
Comment (taylorsw04) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/editableTreeTypes.spec.ts:141]: remove todo?

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/lazyTree.spec.ts [hunk 5]
```diff
@@ -424,6 +449,62 @@ describe("LazyStruct", () => {
...
+	const schemaBuilder = new SchemaBuilder({ scope: "test", libraries: [leafDomain.library] });
+	const structNodeSchema = schemaBuilder.struct("struct", {
+		optional: SchemaBuilder.fieldOptional(leafDomain.string),
+		required: SchemaBuilder.fieldRequired(leafDomain.boolean),
+		sequence: SchemaBuilder.fieldSequence(leafDomain.number),
+	});
+	const schema = schemaBuilder.toDocumentSchema(SchemaBuilder.fieldOptional(Any));
+
+	// Count the number of times edits have been generated.
+	let editCallCount = 0;
```
Comment (taylorsw04) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/lazyTree.spec.ts:475]: unused

**Review by taylorsw04 — CHANGES_REQUESTED** [review]
Some cleanup/opt needed but otherwise great.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyTree.ts [hunk 6]
```diff
@@ -516,34 +521,62 @@ function buildStructClass<TSchema extends StructSchema>(
...
 			},
+			set: setter,
 		};
 
+		// Create set method for property (when appropriate)
+		if (setter !== undefined) {
+			propertyDescriptorMap[`set${capitalize(key)}`] = {
+				enumerable: false,
+				get(this: CustomStruct) {
+					return setters.get(fieldSchema.kind);
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyTree.ts:567]: Nope. Good catch.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts [hunk 7]
```diff
@@ -433,33 +438,40 @@ export type StructTyped<TSchema extends StructSchema> = Struct &
...
-// }
+export type StructFields<TFields extends RestrictiveReadonlyRecord<string, FieldSchema>> = {
+	// boxed fields (TODO: maybe remove these when same as non-boxed version?)
+	readonly [key in keyof TFields as `boxed${Capitalize<key & string>}`]: TypedField<TFields[key]>;
+} & {
+	// Unboxed fields
+	readonly [key in keyof TFields]: UnboxField<TFields[key]>;
+} & {
+	// Add setter (make property writable) when schema type is a value.
+	// If we could map to getters and setters separately, we would preferably do that, but we can't.
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts:458]: This is the TS issue to cite here: https://github.com/microsoft/TypeScript/issues/43826

Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts:458]:
```suggestion
	// If we could map to getters and setters separately, we would preferably do that, but we can't.
	// See https://github.com/microsoft/TypeScript/issues/43826 for more details on this limitation.
```

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts [hunk 8]
```diff
@@ -433,33 +438,40 @@ export type StructTyped<TSchema extends StructSchema> = Struct &
...
-// 	[key in keyof TFields]: UnwrappedField<TFields[key]> & StructSetContent<TFields[key]>;
-// }
+export type StructFields<TFields extends RestrictiveReadonlyRecord<string, FieldSchema>> = {
+	// boxed fields (TODO: maybe remove these when same as non-boxed version?)
+	readonly [key in keyof TFields as `boxed${Capitalize<key & string>}`]: TypedField<TFields[key]>;
+} & {
+	// Unboxed fields
+	readonly [key in keyof TFields]: UnboxField<TFields[key]>;
+} & {
+	// Add setter (make property writable) when schema type is a value.
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts:457]: What does "when schema type is a value" mean here (and below)? Do you mean when the field kind has a logical set operation for the case where the child unboxes to a `Value`?

**Review by taylorsw04 — APPROVED** [review]

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts (hunk 8 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts:457]: Yeah, that's ultimately what I mean. I'll update the comments to be in terms of that.

Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts]: Updated. Should `AssignableFieldKinds` be renamed to `SettableFieldKinds` to mirror this? Or are "assignable" and "settable" reasonably synonymous in this context?

**Review by taylorsw04 — APPROVED** [review]

**Pull request merged**